File: android/app/src/main/kotlin/com/kiya/bankinggenie/dbu/SlmEmbedder.kt
package com.kiya.bankinggenie.dbu

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.nio.LongBuffer
import kotlin.math.sqrt

class SlmEmbedder(private val modelsDir: File) {

    private val TAG = "SlmEmbedder"
    private val vocab: Map<String, Int>
    private val env: OrtEnvironment = OrtEnvironment.getEnvironment()
    private val session: OrtSession

    companion object {
        private const val MAX_SEQ_LEN = 128
        private const val UNK_ID = 100L
        private const val CLS_ID = 101L
        private const val SEP_ID = 102L
        private const val PAD_ID = 0L
    }

    init {
        AppLogger.info(TAG, "Loading vocab from $modelsDir/vocab.txt")
        vocab = loadVocab()
        AppLogger.info(TAG, "Vocab loaded — ${vocab.size} tokens")
        AppLogger.info(TAG, "Loading ONNX model from $modelsDir/all-MiniLM-L12-v2_quantized.onnx")
        val modelBytes = File(modelsDir, "all-MiniLM-L12-v2_quantized.onnx").readBytes()
        session = env.createSession(modelBytes, OrtSession.SessionOptions())
        AppLogger.info(TAG, "ONNX session created")
    }

    fun encode(text: String): FloatArray {
        val tokens = tokenize(text)
        val inputIds      = LongArray(MAX_SEQ_LEN) { PAD_ID }
        val attentionMask = LongArray(MAX_SEQ_LEN) { 0L }
        val tokenTypeIds  = LongArray(MAX_SEQ_LEN) { 0L }

        inputIds[0] = CLS_ID; attentionMask[0] = 1L
        val len = minOf(tokens.size, MAX_SEQ_LEN - 2)
        for (i in 0 until len) { inputIds[i + 1] = tokens[i]; attentionMask[i + 1] = 1L }
        inputIds[len + 1] = SEP_ID; attentionMask[len + 1] = 1L

        val shape = longArrayOf(1, MAX_SEQ_LEN.toLong())
        val tIds  = OnnxTensor.createTensor(env, LongBuffer.wrap(inputIds),      shape)
        val tMask = OnnxTensor.createTensor(env, LongBuffer.wrap(attentionMask), shape)
        val tType = OnnxTensor.createTensor(env, LongBuffer.wrap(tokenTypeIds),  shape)

        val output = session.run(mapOf("input_ids" to tIds, "attention_mask" to tMask, "token_type_ids" to tType))
        @Suppress("UNCHECKED_CAST")
        val hidden = (output.get(0).value as Array<Array<FloatArray>>)[0]

        val embedding = FloatArray(hidden[0].size)
        var count = 0
        for (i in 0 until MAX_SEQ_LEN) {
            if (attentionMask[i] == 1L) {
                val row = hidden[i]
                for (j in embedding.indices) embedding[j] = embedding[j] + row[j]
                count++
            }
        }
        if (count > 0) { val c = count.toFloat(); for (j in embedding.indices) embedding[j] = embedding[j] / c }

        val norm = sqrt(embedding.map { it * it }.sum()).toDouble()
        if (norm > 1e-10) { val n = norm.toFloat(); for (j in embedding.indices) embedding[j] = embedding[j] / n }

        tIds.close(); tMask.close(); tType.close(); output.close()
        return embedding
    }

    fun cosineSimilarity(a: FloatArray, b: FloatArray): Float {
        var dot = 0f
        for (i in a.indices) dot += a[i] * b[i]
        return dot
    }

    private fun tokenize(text: String): LongArray {
        val words = text.lowercase().split(Regex("\\s+|(?=[^a-z0-9])|(?<=[^a-z0-9])")).filter { it.isNotEmpty() }
        val ids = mutableListOf<Long>()
        for (word in words) ids.addAll(wordPieceTokenize(word))
        return ids.toLongArray()
    }

    private fun wordPieceTokenize(word: String): List<Long> {
        if (word in vocab) return listOf(vocab[word]!!.toLong())
        val pieces = mutableListOf<Long>()
        var start = 0
        while (start < word.length) {
            var end = word.length; var found = false
            while (start < end) {
                val sub = if (start == 0) word.substring(start, end) else "##${word.substring(start, end)}"
                if (sub in vocab) { pieces.add(vocab[sub]!!.toLong()); start = end; found = true; break }
                end--
            }
            if (!found) { pieces.add(UNK_ID); break }
        }
        return pieces
    }

    private fun loadVocab(): Map<String, Int> {
        val map = mutableMapOf<String, Int>()
        BufferedReader(InputStreamReader(File(modelsDir, "vocab.txt").inputStream())).use { br ->
            var idx = 0
            br.forEachLine { line -> map[line.trim()] = idx++ }
        }
        return map
    }

    fun close() { session.close(); env.close() }
}

File: ios/Runner/SlmEmbedder.swift
import Foundation

#if canImport(onnxruntime_objc)
import onnxruntime_objc
#endif

private let TAG = "SlmEmbedder"
private let MAX_SEQ_LEN = 128
private let UNK_ID: Int64 = 100
private let CLS_ID: Int64 = 101
private let SEP_ID: Int64 = 102
private let PAD_ID: Int64 = 0

final class SlmEmbedder {

    private let vocab: [String: Int]
    private let log = AppLogger.shared

#if canImport(onnxruntime_objc)
    private let env: ORTEnv
    private let session: ORTSession

    init(modelsDir: URL) throws {
        log.info(TAG, "Loading vocab from \(modelsDir.path)/vocab.txt")
        vocab = try SlmEmbedder.loadVocab(modelsDir: modelsDir)
        log.info(TAG, "Vocab loaded — \(vocab.count) tokens")

        let modelPath = modelsDir.appendingPathComponent("all-MiniLM-L12-v2_quantized.onnx").path
        log.info(TAG, "Loading ONNX model from \(modelPath)")

        env = try ORTEnv(loggingLevel: .warning)
        let opts = try ORTSessionOptions()
        try opts.setIntraOpNumThreads(2)
        session = try ORTSession(env: env, modelPath: modelPath, sessionOptions: opts)
        log.info(TAG, "ONNX session created")
    }

    func encode(_ text: String) throws -> [Float] {
        let tokens = tokenize(text)

        var inputIds      = [Int64](repeating: PAD_ID, count: MAX_SEQ_LEN)
        var attentionMask = [Int64](repeating: 0,      count: MAX_SEQ_LEN)
        var tokenTypeIds  = [Int64](repeating: 0,      count: MAX_SEQ_LEN)

        inputIds[0] = CLS_ID; attentionMask[0] = 1
        let len = min(tokens.count, MAX_SEQ_LEN - 2)
        for i in 0..<len {
            inputIds[i + 1] = tokens[i]
            attentionMask[i + 1] = 1
        }
        inputIds[len + 1] = SEP_ID; attentionMask[len + 1] = 1

        let shape: [NSNumber] = [1, NSNumber(value: MAX_SEQ_LEN)]
        let tIds  = try makeTensor(inputIds,      shape: shape)
        let tMask = try makeTensor(attentionMask, shape: shape)
        let tType = try makeTensor(tokenTypeIds,  shape: shape)

        let inputs: [String: ORTValue] = [
            "input_ids":      tIds,
            "attention_mask": tMask,
            "token_type_ids": tType,
        ]

        let outputs = try session.run(
            withInputs: inputs,
            outputNames: ["last_hidden_state"],
            runOptions: nil
        )

        guard let hiddenValue = outputs["last_hidden_state"] else {
            throw NSError(domain: TAG, code: 1, userInfo: [NSLocalizedDescriptionKey: "Missing last_hidden_state output"])
        }

        let tensorData = try hiddenValue.tensorData() as Data
        let hiddenSize = tensorData.count / MemoryLayout<Float>.size / MAX_SEQ_LEN
        var floats = [Float](repeating: 0, count: MAX_SEQ_LEN * hiddenSize)
        _ = floats.withUnsafeMutableBytes { tensorData.copyBytes(to: $0) }

        var embedding = [Float](repeating: 0, count: hiddenSize)
        var count = 0
        for i in 0..<MAX_SEQ_LEN {
            guard attentionMask[i] == 1 else { continue }
            let offset = i * hiddenSize
            for j in 0..<hiddenSize { embedding[j] += floats[offset + j] }
            count += 1
        }
        if count > 0 {
            let cf = Float(count)
            for j in 0..<hiddenSize { embedding[j] /= cf }
        }

        let norm = sqrt(embedding.reduce(0) { $0 + $1 * $1 })
        if norm > 1e-10 {
            for j in 0..<hiddenSize { embedding[j] /= norm }
        }
        return embedding
    }

    private func makeTensor(_ ids: [Int64], shape: [NSNumber]) throws -> ORTValue {
        let data = ids.withUnsafeBytes { Data($0) }
        return try ORTValue(
            tensorData: NSMutableData(data: data),
            elementType: .int64,
            shape: shape
        )
    }

#else
    // ONNX runtime not available — stub for compilation only
    init(modelsDir: URL) throws {
        log.warn(TAG, "onnxruntime_objc not available — SlmEmbedder disabled")
        vocab = try SlmEmbedder.loadVocab(modelsDir: modelsDir)
    }

    func encode(_ text: String) throws -> [Float] {
        throw NSError(domain: TAG, code: 99, userInfo: [NSLocalizedDescriptionKey: "ONNX runtime not available"])
    }
#endif

    func cosineSimilarity(_ a: [Float], _ b: [Float]) -> Float {
        var dot: Float = 0
        for i in 0..<min(a.count, b.count) { dot += a[i] * b[i] }
        return dot
    }

    func close() {}

    // MARK: - Tokenisation (BERT WordPiece)

    private func tokenize(_ text: String) -> [Int64] {
        let lower = text.lowercased()
        var words: [String] = []
        var current = ""
        for ch in lower {
            if ch.isLetter || ch.isNumber {
                current.append(ch)
            } else {
                if !current.isEmpty { words.append(current); current = "" }
                let s = String(ch)
                if !s.trimmingCharacters(in: .whitespaces).isEmpty { words.append(s) }
            }
        }
        if !current.isEmpty { words.append(current) }
        var ids: [Int64] = []
        for word in words { ids.append(contentsOf: wordPieceTokenize(word)) }
        return ids
    }

    private func wordPieceTokenize(_ word: String) -> [Int64] {
        if let id = vocab[word] { return [Int64(id)] }
        var pieces: [Int64] = []
        var start = word.startIndex
        while start < word.endIndex {
            var end = word.endIndex
            var found = false
            while start < end {
                let sub = start == word.startIndex
                    ? String(word[start..<end])
                    : "##" + String(word[start..<end])
                if let id = vocab[sub] {
                    pieces.append(Int64(id))
                    start = end
                    found = true
                    break
                }
                end = word.index(before: end)
            }
            if !found { return [UNK_ID] }
        }
        return pieces
    }

    private static func loadVocab(modelsDir: URL) throws -> [String: Int] {
        let vocabPath = modelsDir.appendingPathComponent("vocab.txt")
        let content = try String(contentsOf: vocabPath, encoding: .utf8)
        var map: [String: Int] = [:]
        var idx = 0
        for line in content.components(separatedBy: "\n") {
            let token = line.trimmingCharacters(in: .whitespacesAndNewlines)
            if !token.isEmpty { map[token] = idx }
            idx += 1
        }
        return map
    }
}
